### PR TITLE
Implemented unit tests for Book model

### PR DESCRIPTION
**Pull request by tagonzalez**
## Summary 
 [Crear tests para el modelo de Book](https://trello.com/c/eibnrVHu/20-crear-tests-para-el-modelo-de-book)

Added unit tests for Book model to test the following functionality:
- Validity of a newly created book
- Validation of the required fields according to the  [Crear modelo de Book Trello Card](https://trello.com/c/P7v1zr8q/18-crear-modelo-de-book)

## Notes
For the `validate_presence_of` method to work, it was necessary to install the `shoulda-matchers` gem.

## Screenshots
book-tests-screenshot

### CODE REVIEW

File: spec/models/book_spec.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+describe Book do
+  subject(:book) { FactoryBot.create(:book) }
+
+  it { is_expected.to be_valid }
+  it { is_expected.to validate_presence_of(:genre) }
+  it { is_expected.to validate_presence_of(:author) }
+  it { is_expected.to validate_presence_of(:image) }
+  it { is_expected.to validate_presence_of(:title) }
+  it { is_expected.to validate_presence_of(:editor) }
+  it { is_expected.to validate_presence_of(:year) }
```
Comment (mcolombo87) [inline spec/models/book_spec.rb]: I thinks that you must validate if an book can by saved correctly in database, right now you are only testing the models validators. What do you think?

Comment (tagonzalez) [reply spec/models/book_spec.rb]: I think that that's already tested by calling `FactoryBot.create(:book)` in line 6 which returns an instance of a Book which has already been saved to the DB according to the [FactoryBot docs](https://www.rubydoc.info/gems/factory_bot/file/GETTING_STARTED.md#Using_factories).

If I'm not mistaken, this means that by passing the test in line 8, the Book instance created on line 6 is valid and has been saved correctly to the DB.

File: Gemfile [hunk 2]
```diff
@@ -38,9 +38,14 @@ group :development, :test do
...
   # Adds support for Capybara system testing and selenium driver
   gem 'capybara', '>= 2.15'
+  gem 'database_cleaner', '~> 1.7.0'
+  gem 'factory_bot', '~> 5.0.2'
+  gem 'faker', '~> 1.9.6'
+  gem 'rspec-rails', '~> 3.8'
   gem 'rubocop', '~> 0.73.0'
   gem 'rubocop-rails', '~> 2.2.1'
   gem 'selenium-webdriver'
+  gem 'shoulda-matchers', '~> 4.1.1', require: false
```
Comment (marcegl91) [inline Gemfile]: Why did you set require to false?

File: spec/models/book_spec.rb (hunk 1 above)
Comment (marcegl91) [reply spec/models/book_spec.rb]: I'd prefer to build the model (like the example given in the card) instead of create (because it is not persisted) and then use the assertions you made. In another test, you can create the model and then test if it is persisted (which is another thing different from being valid).

Also, you are not testing failure cases for this model, validations could be broken and you will never know that because you are only testing a happy path.

File: .travis.yml [hunk 3]
```diff
@@ -17,13 +17,14 @@ before_install:
   - gem install bundler
 
 before_script:
-  - psql -c "CREATE USER \"rails-bootstrap\" WITH PASSWORD 'rails-bootstrap';" -U postgres
-  - psql -c "DROP DATABASE IF EXISTS \"rails-bootstrap_test\";" -U postgres
-  - psql -c "CREATE DATABASE \"rails-bootstrap_test\" ENCODING 'utf8';" -U postgres
+  - psql -c "CREATE USER \"wbooks-api\" WITH PASSWORD 'wbooks-api';" -U postgres
```
Comment (marcegl91) [inline .travis.yml]: These changes are already in the master branch, why are they here in this PR?

File: Gemfile (hunk 2 above)
Comment (tagonzalez) [reply Gemfile]: I was following what it now seems to be outdated documentation to install `shoulda-matchers`.
The flag will be removed in the next commits.

File: .travis.yml (hunk 3 above)
Comment (tagonzalez) [reply .travis.yml]: These changes made it to branch `master` because of a requested change PR #3 

Before those requested changes were made in #3 , I had already started work in a new branch that kept the old `.travis.yml` file and that didn't cause any issues until this PR were I'm actually using the `test` database to run the tests. I had to make the changes for the tests to work.

Is there a better/preffered way to deal with these situations?

File: spec/models/book_spec.rb (hunk 1 above)
Comment (tagonzalez) [reply spec/models/book_spec.rb]: I agree that unhappy path tests are missing.
On further reading I learned about the difference between the model being valid and it being saved and changed the `create` method for `build`. 
This change and additional failure cases tests can be found in the latest commits.

File: .travis.yml (hunk 3 above)
Comment (marcegl91) [reply .travis.yml]: Maybe you want to change the base branch of this PR to the new PR #8, then review the "files changed" tab on the PR prior to saving in order to check that everything is ok. If it messes up even more, leave this way, because you eventually are going to change the base branch to master.

Anyway, you can wait until everything is merged to master and then when you change the base branch to master, the "files changed" tab will reflect only the changes that are different from master and nothing else.

File: spec/models/book_spec.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe Book do
+  required_fields = Book.validators.first.attributes
```
Comment (marcegl91) [inline spec/models/book_spec.rb]: This is really a great idea, I've never thought about doing that but I'm deeply saddened to inform you that this will not work.

Let's imagine the following scenario:

1) You have a model User with a `validates :password, presence: true`
2) Then, you write a test like this one in order to test that the validations are there, and these changes go to master.
3) I'm a careless guy so without noticing it, I delete the validation of user's password in the user.rb file.
4) I run the test suite and now the password validator is not in Book.validators list, so the test suite will return that everything is ok.
5) I merge that to master and no one realizes that now our users can sign up without a password.

Trying to write simple, DRY code is something that is valued in this community so I like that you tried to do this way, excellent work! :100:

Sadly, I can not approve this PR because of that detail.

File: spec/models/book_spec.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+  context 'with correctly typed required fields' do
+    subject { FactoryBot.build(:book) }
```
Comment (marcegl91) [inline spec/models/book_spec.rb]: I think that doing build(:book) should work the same way, like the card example.

File: spec/models/book_spec.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    it { is_expected.to be_valid }
+
+    required_fields.each do |field|
+      it { is_expected.to validate_presence_of(field) }
+    end
+
+    it 'should save correctly' do
```
Comment (marcegl91) [inline spec/models/book_spec.rb]: I think this works so leave this way. But I prefer creating a new context and doing something like this:

```
  context 'when a valid sapo_pepe is created' do
    subject(:sapo_pepe) { create(:sapo_pepe) }

    it { is_expected.to be_persisted }
  end
```
tl;dr: its ok :)

File: spec/models/book_spec.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      expect(subject.save).to be true
+    end
+  end
+
+  required_fields.each do |field|
```
Comment (marcegl91) [inline spec/models/book_spec.rb]: This is the same thing as before. Now :password validator is not in that list so it won't test if the password is there.

File: .travis.yml (hunk 3 above)
Comment (tagonzalez) [reply .travis.yml]: Seems like you can't preview changing the base before actually changing it, it looks like there it made no difference to change it to the new PR #8 . I'll leave it like it is now, comparing to #8

File: spec/models/book_spec.rb (hunk 4 above)
Comment (tagonzalez) [reply spec/models/book_spec.rb]: I understand the problem you're mentioning and it makes sense.
If I understand correctly, it's just a matter of hardcoding the `required_fields` variable to an array of the fields and if eventually a new validation is added, remember to add it in the code later.

Should be solved with
```  required_fields = %i[genre author image title editor year]```
as done in the latest commit.

File: spec/models/book_spec.rb (hunk 5 above)
Comment (tagonzalez) [reply spec/models/book_spec.rb]: You're right! 
Changes were committed and it's much cleaner now :)

File: spec/support/factories/book.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :book do
+    genre { Faker::Book.genre }
+    author { Faker::Book.author }
+    image { Faker::LoremPixel.image }
```
Comment (marcegl91) [inline spec/support/factories/book.rb]: I didn't know about this Faker module, this is good! :+1:
